core/shared/src/main/scala/zio/ZSchedule.scala

```diff
@@ -378,82 +395,108 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
+   *
+   * Note that this does not apply to sleeps performed in ZSchedule#initial.
+   * All effects executed while calculating the modified duration will run with the old
+   * environment.
+   */
+  final def modifyDelay_[R1](
+    f: (B, Duration) => ZIO[R1, Nothing, Duration]
+  )(
+    g: (Clock.Service[Any] => Clock.Service[Any]) => R1 => R
+  ): ZSchedule[R1 with Clock, A, B] = {
+    def proxy(clock0: Clock.Service[Any], env: R1, current: B): Clock.Service[Any] = new Clock.Service[Any] {
+      def currentTime(unit: TimeUnit) = clock0.currentTime(unit)
+      def currentDateTime             = clock0.currentDateTime
+      val nanoTime                    = clock0.nanoTime
+      def sleep(duration: Duration)   = f(current, duration).provide(env).flatMap(clock0.sleep)
+    }
+    new ZSchedule[R1 with Clock, A, B] {
+      type State = self.State
+      val initial = self.initial.provideSome[R1](g(identity))
+      val extract = (a: A, s: self.State) => self.extract(a, s)
+      val update = (a: A, s: self.State) =>
+        self.update(a, s).provideSome[R1](env => g(proxy(_, env, self.extract(a, s)))(env))
+    }
+  }
 
   /**
    * A new schedule that applies the current one but runs the specified effect
    * for every decision of this schedule. This can be used to create schedules
    * that log failures, decisions, or computed values.
    */
-  final def onDecision[A1 <: A](f: (A1, ZSchedule.Decision[State, B]) => UIO[Unit]): ZSchedule[R, A1, B] =
+  final def onDecision[A1 <: A](f: (A1, self.State) => UIO[Unit]): ZSchedule[R, A1, B] =
     updated(update => (a, s) => update(a, s).tap(step => f(a, step)))
 
   /**
-   * Returns a new schedule that reconsiders the decision made by this schedule.
+   * Provide all requirements to the schedule.
    */
-  final def reconsider[A1 <: A, C](
-    f: (A1, ZSchedule.Decision[State, B]) => ZSchedule.Decision[State, C]
-  ): ZSchedule[R, A1, C] =
-    reconsiderM((a, s) => IO.succeed(f(a, s)))
+  final def provide(r: R): ZSchedule[Any, A, B] =
+    provideSome(_ => r)
+
+  /**
+   * Provide some of the requirements to the schedule.
+   */
+  final def provideSome[R1](f: R1 => R): ZSchedule[R1, A, B] =
+    new ZSchedule[R1, A, B] {
+      type State = self.State
+      val initial = self.initial.provideSome(f)
+      val extract = self.extract
+      val update  = (a: A, s: self.State) => self.update(a, s).provideSome(f)
+    }
 
   /**
    * Returns a new schedule that effectfully reconsiders the decision made by
    * this schedule.
+   * The provided either will be a Left if the schedule has failed and will contain the old state
+   * or a Right with the new state if the schedule has updated successfully.
    */
-  final def reconsiderM[A1 <: A, C](
-    f: (A1, ZSchedule.Decision[State, B]) => UIO[ZSchedule.Decision[State, C]]
-  ): ZSchedule[R, A1, C] =
+  final def reconsider[R1 <: R, A1 <: A](
+    f: (A1, Either[State, State]) => ZIO[R1, Unit, State]
```

Comment:
This isn't quite equivalent to the old reconsider, nor can it be without fixing the env.

Reply:
I'm not quite sure what we are losing compared to the old reconsider.
Can you explain in a little more detail?